### PR TITLE
Respect and pass through `ANDROID_USER_HOME` and `ANDROID_AVD_HOME`

### LINKED ISSUE #1665: Cannot support non-standard location for `.android` directory
### Describe the bug

The default location for the Android SDK's `.android` directory is `$HOME/.android` but it can be an arbitrary directory via `ANDROID_USER_HOME`. When Briefcase creates an AVD, it tries to modify its configuration afterwards; this fails because it assumes the AVD is located within `$HOME/.android`. More generally, enumerating the list of AVDs assumes this location as well.

### Steps to reproduce

Run `ANDROID_USER_HOME=/tmp/android briefcase run android` and create a new emulator.

```
❯ ANDROID_USER_HOME=/tmp/android briefcase run android -vv
Using JDK at /home/russell/.cache/briefcase/tools/java17
Using Android SDK at /home/russell/.cache/briefcase/tools/android_sdk
Android emulator is already installed.

>>> Running Command:
>>>     /home/russell/.cache/briefcase/tools/android_sdk/platform-tools/adb devices -l
>>> Working Directory:
>>>     /home/russell/tmp/beeware/helloworld
>>> Command Output:
>>>     List of devices attached
>>>     
>>> Return code: 0

>>> Running Command:
>>>     /home/russell/.cache/briefcase/tools/android_sdk/emulator/emulator -list-avds
>>> Working Directory:
>>>     /home/russell/tmp/beeware/helloworld
>>> Command Output:
>>>     beePhone
>>> Return code: 0

Select device:

  1) @beePhone (emulator)
  2) Create a new Android emulator

> 2

>>> Running Command:
>>>     /home/russell/.cache/briefcase/tools/android_sdk/emulator/emulator -list-avds
>>> Working Directory:
>>>     /home/russell/tmp/beeware/helloworld
>>> Command Output:
>>>     beePhone
>>> Return code: 0

You need to select a name for your new emulator. This is an identifier that
can be used to start the emulator in future. It should follow the same naming
conventions as a Python package (i.e., it may only contain letters, numbers,
hyphens and underscores). If you don't provide a name, Briefcase will use the
a default name 'beePhone2'.

Emulator name [beePhone2]: 
Device skin 'pixel_7_pro' already exists.

>>> Running Command:
>>>     /home/russell/.cache/briefcase/tools/android_sdk/cmdline-tools/9.0/bin/avdmanager --verbose create avd --name beePhone2 --abi x86_64 --package 'system-images;android-31;default;x86_64' --device pixel
>>> Working Directory:
>>>     /home/russell/tmp/beeware/helloworld
>>> Environment Overrides:
>>>     ANDROID_HOME=/home/russell/.cache/briefcase/tools/android_sdk
>>>     ANDROID_SDK_ROOT=/home/russell/.cache/briefcase/tools/android_sdk
>>>     JAVA_HOME=/home/russell/.cache/briefcase/tools/java17
>>> Command Output:
>>>     Loading local repository...                                                     
>>>     [=========                              ] 25% Loading local repository...       
>>>     [=========                              ] 25% Fetch remote repository...        
>>>     [=======================================] 100% Fetch remote repository...       
>>> Return code: 0
Creating Android emulator beePhone2... done
Adding extra device configuration... errored

Unable to read configuration of AVD @beePhone2

```

### Expected behavior

Briefcase respects `ANDROID_USER_HOME` or similar.

### Screenshots

_No response_

### Environment

- Operating System: pop os 22.04
- Python version: 3.10
- Software versions:
  - Briefcase: `0.3.18.dev72+ge141754e`

### Logs

[briefcase.2024_02_24-16_41_28.run.log](https://github.com/beeware/briefcase/files/14394489/briefcase.2024_02_24-16_41_28.run.log)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rmartin16**
## Changes
- Allows Briefcase to use `ANDROID_USER_HOME` as the location of the `.android` directory and `ANDROID_AVD_HOME` as the location of the AVDs; these were previously hardcorded to their defaults, `$HOME/.android` and `$HOME/.android/avd`
- Closes #1665

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [X] I have read the **CONTRIBUTING.md** file
- [X] I will abide by the code of conduct

### CODE REVIEW

**Comment by rmartin16** [comment]
Unfortunately, the emulator makes a bit of a mess of all this.

## tl;dr
- If the environment variables don't point to already valid directories, the emulator just ignores them and falls back to the default directories.
  - This is especially problematic if `~/.android` already contains AVDs because `emulator -list-avds` will return AVDs from `~/.android` but there isn't any context to know the `ANDROID_USER/AVD_HOME` env vars are being ignored
  - So, when these env vars are set and Briefcase tries to start an emulator from `~/.android`, things aren't going to go well

## Environment
### Assume this env:
```console
❯ env | grep ANDROID
ANDROID_HOME=/opt/sdks/Android/sdk/
```

### Assume these directory layouts:
<details>
<summary>Normal <code>~/.android</code> for an AVD called <code>android-home-dir</code></summary>

```console
❯ tree -d ~/.android
/home/russell/.android
├── avd
│   └── android-home-dir.avd
│       ├── modem_simulator
│       │   └── etc
│       │       └── modem_simulator
│       │           └── files
│       ├── snapshots
│       │   └── default_boot
│       └── tmpAdbCmds
└── cache
```
</details>

<details>
<summary><code>/tmp/android-user-home/</code> is a copy of <code>~/.android</code> for an AVD called <code>android-user-home</code></summary>

```console
❯ tree -d /tmp/android-user-home/
/tmp/android-user-home/
├── avd
│   └── android-user-home.avd
│       ├── modem_simulator
│       │   └── etc
│       │       └── modem_simulator
│       │           └── files
│       ├── snapshots
│       │   └── default_boot
│       └── tmpAdbCmds
└── cache
```
</details>

<details>
<summary><code>/tmp/android-avd-home/</code> which is just the <code>avd</code> directory of <code>~/.android</code> for an AVD called <code>avd-home</code></summary>

```console
❯ tree -d /tmp/android-avd-home/
/tmp/android-avd-home/
└── avd-home.avd
    ├── modem_simulator
    │   └── etc
    │       └── modem_simulator
    │           └── files
    ├── snapshots
    │   └── default_boot
    └── tmpAdbCmds
```
</details>

## Consider `emulator`
Using the most recently released version of the emulator
```console
❯ /opt/sdks/Android/sdk/emulator/emulator -version
INFO    | Android emulator version 33.1.24.0 (build_id 11237101) (CL:N/A)
```
Correctly finds the AVD in `~/.android/`
```console
❯ /opt/sdks/Android/sdk/emulator/emulator -list-avds
android-home-dir
```
Respects `ANDROID_AVD_HOME` and finds the AVD there
```console
❯ ANDROID_AVD_HOME=/tmp/android-avd-home/ /opt/sdks/Android/sdk/emulator/emulator -list-avds
avd-home
```
Ignores `ANDROID_USER_HOME` and finds the AVD in `~/.android/` (!!!!)
```console
❯ ANDROID_USER_HOME=/tmp/android-user-home/ /opt/sdks/Android/sdk/emulator/emulator -list-avds
android-home-dir
```
For shits and giggles.....I duplicated `/tmp/android-user-home` to `/tmp/.android` and apparently `emulator` does respect `ANDROID_SDK_HOME`....so there's that...
```console
❯ ANDROID_SDK_HOME=/tmp/ /opt/sdks/Android/sdk/emulator/emulator -list-avds
android-user-home
```
but `ANDROID_HOME` is not respected in the same way
```console
❯ ANDROID_HOME=/tmp/.android/ /opt/sdks/Android/sdk/emulator/emulator -list-avds
android-home-dir
```
Even more fascinatingly, `emulator` apparently doesn't use `ANDROID_EMULATOR_HOME` to find AVDs...it just uses it for whatever other data its putting in there...
```console
❯ ANDROID_EMULATOR_HOME=/tmp/android-user-home/ /opt/sdks/Android/sdk/emulator/emulator -list-avds
android-home-dir
```

## Consider `avdmanager`
Without env vars, creates the AVD in `~/.android`
```console
❯ /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager --verbose create avd --name android-no-env-vars --abi x86_64 --package 'system-images;android-31;default;x86_64' --device pixel
[=======================================] 100% Fetch remote repository...       

❯ tree -d ~/.android/
/home/russell/.android/
├── avd
│   ├── android-home-dir.avd
│   │   ├── modem_simulator
│   │   │   └── etc
│   │   │       └── modem_simulator
│   │   │           └── files
│   │   ├── snapshots
│   │   │   └── default_boot
│   │   └── tmpAdbCmds
│   └── android-no-env-vars.avd
└── cache
```
Respects `ANDROID_USER_HOME` and creates AVD in `/tmp/android-user-home`
```console
❯ ANDROID_USER_HOME=/tmp/android-user-home/ /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager --verbose create avd --name android-user-home-env-var --abi x86_64 --package 'system-images;android-31;default;x86_64' --device pixel
[=======================================] 100% Fetch remote repository...

❯ tree -d /tmp/android-user-home/
/tmp/android-user-home/
├── avd
│   ├── android-user-home.avd
│   │   ├── modem_simulator
│   │   │   └── etc
│   │   │       └── modem_simulator
│   │   │           └── files
│   │   ├── snapshots
│   │   │   └── default_boot
│   │   └── tmpAdbCmds
│   └── android-user-home-env-var.avd
└── cache
```
Respects `ANDROID_AVD_HOME` and creates AVD in `/tmp/android-avd-home`
```console
❯ ANDROID_AVD_HOME=/tmp/android-avd-home/ /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager --verbose create avd --name android-avd-home-env-var --abi x86_64 --package 'system-images;android-31;default;x86_64' --device pixel
[=======================================] 100% Fetch remote repository...       

❯ tree -d /tmp/android-avd-home/
/tmp/android-avd-home/
├── android-avd-home-env-var.avd
└── avd-home.avd
    ├── modem_simulator
    │   └── etc
    │       └── modem_simulator
    │           └── files
    ├── snapshots
    │   └── default_boot
    └── tmpAdbCmds
```
Allows `ANDROID_AVD_HOME` to override `ANDROID_USER_HOME`
```console
❯ ANDROID_USER_HOME=/tmp/android-user-home/ ANDROID_AVD_HOME=/tmp/android-avd-home/ /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager --verbose create avd --name android-avd-both-env-var --abi x86_64 --package 'system-images;android-31;default;x86_64' --device pixel
[=======================================] 100% Fetch remote repository...       

❯ tree -d /tmp/android-user-home/
/tmp/android-user-home/
├── avd
│   ├── android-user-home.avd
│   │   ├── modem_simulator
│   │   │   └── etc
│   │   │       └── modem_simulator
│   │   │           └── files
│   │   ├── snapshots
│   │   │   └── default_boot
│   │   └── tmpAdbCmds
│   └── android-user-home-env-var.avd
└── cache

11 directories

❯ tree -d /tmp/android-avd-home/
/tmp/android-avd-home/
├── android-avd-both-env-var.avd
├── android-avd-home-env-var.avd
└── avd-home.avd
    ├── modem_simulator
    │   └── etc
    │       └── modem_simulator
    │           └── files
    ├── snapshots
    │   └── default_boot
    └── tmpAdbCmds
```

**Comment by rmartin16** [comment]
Briefcase is already [assuming](https://github.com/beeware/briefcase/blob/35a2e37ec4363f74ad4ee1b477bad0c1bae295ea/src/briefcase/integrations/android_sdk.py#L131) some of the internal structure of the `avd` directory; so, if we extend that pattern here, we could inspect the filesystem while emulators are being enumerated to disqualify ones that don't exist where we're expecting.

Specifically, while iterating the output from `emulator -list-avds`, use the emulator name to check if `self.avd_path / f"{avd}.avd` is, in fact, a directory.

This would allow Briefcase to tolerate the emulator's fallback behavior; although, all this is likely to be confusing to users who have `ANDROID_AVD_HOME` set currently but have been using Briefcase with the `~/.android/avd` directory. After the upgrade, their AVDs won't be listed anymore...

**Comment by rmartin16** [comment]
Actually....maybe the answer is using `avdmanager` instead of `emulator` to find existing AVDs:
```console
❯ /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact
android-home-dir
android-no-env-vars

❯ ANDROID_USER_HOME=/tmp/android-user-home /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact
android-user-home
android-user-home-env-var

❯ ANDROID_USER_HOME=/tmp/android-user-home-none /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact

❯ ANDROID_AVD_HOME=/tmp/android-avd-home /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact
android-avd-both-env-var
android-avd-home-env-var

❯ ANDROID_AVD_HOME=/tmp/android-avd-home-none /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact
android-home-dir
android-no-env-vars

❯ ANDROID_USER_HOME=/tmp/android-user-home-none ANDROID_AVD_HOME=/tmp/android-avd-home-none /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact

❯ ANDROID_USER_HOME=/tmp/android-user-home-none ANDROID_AVD_HOME=/tmp/android-avd-home /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact
android-avd-both-env-var
android-avd-home-env-var

❯ ANDROID_USER_HOME=/tmp/android-user-home ANDROID_AVD_HOME=/tmp/android-avd-home-none /opt/sdks/Android/sdk/cmdline-tools/9.0/bin/avdmanager list avd --compact
```

Notably:
- `avdmanager` will use `ANDROID_AVD_HOME` if the directory exists...otherwise, it falls back to the defaults
- `avdmanager` will just create `ANDROID_USER_HOME` if the directory doesn't exist...and not fall back to defaults
- `ANDROID_AVD_HOME` appears to take precedence over `ANDROID_USER_HOME`

This seems like much more sane behavior....although, we'll want to check if `ANDROID_USER_HOME` doesn't exist yet to avoid `avdmanager` eagerly creating it.

Furthermore, I'll need to explore the behavior around an already running emulator that doesn't exist inside either of the directories from these env vars....ideally, such a running emulator would still be usable by Briefcase.

**Comment by freakboy3742** [comment]
> Actually....maybe the answer is using `avdmanager` instead of `emulator` to find existing AVDs:
...
> Notably:
> 
> * `avdmanager` will use `ANDROID_AVD_HOME` if the directory exists...otherwise, it falls back to the defaults
> * `avdmanager` will just create `ANDROID_USER_HOME` if the directory doesn't exist...and not fall back to defaults
> * `ANDROID_AVD_HOME` appears to take precedence over `ANDROID_USER_HOME`
> 
> This seems like much more sane behavior....although, we'll want to check if `ANDROID_USER_HOME` doesn't exist yet to avoid `avdmanager` eagerly creating it.

Agreed - the only point of concern I have is the "fallback to defaults". It's easy enough to make a typo in a directory name, think you've set an environment variable and then discover your writing to a default location. On the one hand, this *is* the behavior of Android itself, so maybe we shouldn't fight it too hard; but on the other hand, we're in a position to at least *validate* if the value is going to cause a fallback to defaults and warn the user.

**Comment by rmartin16** [comment]
> Agreed - the only point of concern I have is the "fallback to defaults". It's easy enough to make a typo in a directory name, think you've set an environment variable and then discover your writing to a default location. On the one hand, this _is_ the behavior of Android itself, so maybe we shouldn't fight it too hard; but on the other hand, we're in a position to at least _validate_ if the value is going to cause a fallback to defaults and warn the user.

So, first, it is clear that no one is putting together QA plans to ensure these tools don't have nasty corner cases where behavior is completely surprising 😤

Second....I think we can avoid the use of default directories for `avdmanager` despite this.

If a user sets `ANDROID_HOME_USER`, then this all works as expected because `avdmanager` always respects it.

If a user sets only `ANDROID_AVD_HOME`, then `avdmanager` will respect it if we just ensure the directory exists first.

In either case, we should also avoid listing AVDs with `avdmanager` if the directory doesn't exist...because then `avdmanager` will create it.

I think this is the best we can do; then, if there's a typo but AVDs exist in the "real" directory, we won't create the typo directory and the user won't see any of their AVDs listed....prompting investigation. Alternatively, if the directory has a typo and the user attempts to create an AVD with Briefcase, well, the typo directory comes to life.

File: src/briefcase/integrations/android_sdk.py [hunk 1]
```diff
@@ -557,7 +566,7 @@ def list_packages(self):
         try:
             # check_output always writes its output to debug
             self.tools.subprocess.check_output(
-                [os.fsdecode(self.sdkmanager_path), "--list_installed"],
```
Comment (rmartin16) [inline src/briefcase/integrations/android_sdk.py:560]: Removed `os.fsdecode()` calls for `subprocess` args since it has built-in `fspath` support.

File: tests/integrations/android_sdk/AndroidSDK/test__create_emulator.py [hunk 2]
```diff
@@ -184,6 +183,69 @@ def test_create_emulator_with_defaults(
     assert "skin.name=pixel_7_pro" in config
 
 
+def test_create_emulator_with_nonexistent_avd_home(
```
Comment (rmartin16) [inline tests/integrations/android_sdk/AndroidSDK/test__create_emulator.py:186]: I spent way too long trying to add an ability to check if `mkdir()` was called in the existing tests...and nothing was working....so, there's this test too now.

**Comment by freakboy3742** [comment]
@rmartin16 To confirm - is this ready for final review?

**Comment by rmartin16** [comment]
Yep. I've been trying to keep all my PRs as draft until they are ready for review/merge.

**Review by freakboy3742 — APPROVED** [review]
This all looks good to me!
